src/allocator.rs

```diff
@@ -0,0 +1,49 @@
+use std::{
+    alloc::{AllocErr, AllocRef, GlobalAlloc, Layout},
+    ffi::c_void,
+    ptr::NonNull,
+};
+
+use crate::boehm;
+
+pub struct GlobalAllocator;
+pub struct GcAllocator;
+
+unsafe impl GlobalAlloc for GlobalAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        boehm::GC_malloc_uncollectable(layout.size()) as *mut u8
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, _: Layout) {
+        boehm::GC_free(ptr as *mut c_void);
+    }
+
+    unsafe fn realloc(&self, ptr: *mut u8, _: Layout, new_size: usize) -> *mut u8 {
+        boehm::GC_realloc(ptr as *mut c_void, new_size) as *mut u8
+    }
+}
+
+unsafe impl AllocRef for GcAllocator {
+    unsafe fn alloc(&mut self, layout: Layout) -> Result<(NonNull<u8>, usize), AllocErr> {
+        boehm::GC_gcollect();
+        Ok((
+            NonNull::new_unchecked(boehm::GC_malloc(layout.size()) as *mut u8),
```

Comment:
Can `GC_malloc` return `NULL` or does it simply abort?

Reply:
That's a good question. The answer is it's not entirely clear, Boehm docs state that it tries as best it can to emulate ordinary `malloc` calls, so based on this I'm assuming yes it is possible. For this reason I've added an assertion here: 7261bfa41d4a624b101280d03605fe50480aa13d